### ISSUE TITLE
Support for Zoom

### ISSUE THREAD

**Issue by tquizzle** [bdkjones/MuzzleApp#62]
I use Zoom for work and would like Muzzle to work with Zoom. It doesn't seem to do so.

Mac: 10.14.6

**Comment by jdgiotta** [comment]
@tquizzle doesn't work as advertise?

**Comment by cpinto** [comment]
@tquizzle I use Zoom daily and can confirm that Muzzle works as advertised.

**Comment by jspiro** [comment]
What OS? It doesn’t work for zoom on OS X 10.14 latest. It used to work and stopped, maybe after they had their security incident and Apple blocked some of their software. I can confirm it hasn’t worked at all for me recently.

**Comment by dan2k3k4** [comment]
@jspiro it works on latest macOS Catalina 10.15.1 Beta (19B77a).

Although I had to uninstall and then reinstall to get the pop-up about the new security settings that Apple checks for now 🙄.

**Comment by jspiro** [comment]
Not sure what to say 🤷‍♂ I'm running 10.15.1 and 10.15.2 and I use Zoom for hours each day, and notifications are not muzzled at all. I enable it manually.

Is there a commandline API I can call that would return whether I'm in a situation Muzzle should be active during, or some other way?

**Comment by bdkjones** [comment]
The most likely cause is that Muzzle doesn’t have permission to use System Events. Check your settings in System Preferences.

**Comment by dan2k3k4** [comment]
@jspiro open up `System Preferences` then head to `Security & Privacy`, scroll down to `Automation` and you should see `Muzzle` in the list with `System Events` checked:
![image](https://user-images.githubusercontent.com/158704/68661199-d485c500-053a-11ea-9df4-edf2cd252255.png)

If it's not there, then you might need to reinstall the app and then run the `Mach Override` option in the app preferences:
![image](https://user-images.githubusercontent.com/158704/68661364-23335f00-053b-11ea-9caa-6ae059c6fd98.png)

**Comment by bdkjones** [comment]
Ha, the Mach Override does nothing. But I support telling people to click it.

**Comment by jspiro** [comment]
@dan2k3k4 Yeah, Muzzle is not at all in Automation. I've tried a few things to coax it in there, but the permission won't come up and I can't drag it in. Experienced this in Catalina 10.15.2 Beta (19C32e) but also 10.15.1 RC.

Anything I can do to force the permission request?

**Comment by dan2k3k4** [comment]
@jspiro from https://macosxautomation.com/automator/security.html it looks like you could use `tccutil reset AppName` so perhaps `tccutil reset Muzzle` or it might need to be the full path like `tccutil reset /Applications/Muzzle.app`

Another site with info on `tccutil`: https://bitsplitting.org/2018/07/11/reauthorizing-automation-in-mojave/

And perhaps this StackExchange answer (or other answers) might help: https://apple.stackexchange.com/questions/339078/cannot-add-any-apps-to-security-privacy-preferences-on-mojave

**Comment by tleish** [comment]
I experience the same issue.  I installed Muzzle this last week and while presenting in Zoom I received notifications, which led me here.

After Muzzle installed I enabled it under Security & Privacy > **Accessibility**, but I do not see it listed under Security & Privacy > **Automation**.  I have other apps listed there, just not Muzzle.  I tried uninstalling muzzle and re-installing, but did not fix it.  After install, the app only requests permission for Accessibility.

<img width="415" alt="muzzle" src="https://user-images.githubusercontent.com/1099993/75928251-3c825300-5e2b-11ea-9b8b-3facca6cbb53.png">

After running Zoom and sharing screen, I'm never asked for automation access for Muzzle.

MacOS Mojave 10.14.6

**Comment by bdkjones** [comment]
Zoom may not be supported. Muzzle can only work with apps that use Apple's "modern" (released in 2008) screensharing API. Lots of other apps work by simply snapping 30 screenshots per second and stitching those into a video. You can tell which apps those are because they'll spike your CPU while active. The official API sends data directly from the screen buffer as the screen is updated, which is far, far more efficient.

**Comment by tleish** [comment]
So, your thinking it has nothing to do with Muzzle not listed under System Events?

**Comment by dan2k3k4** [comment]
I started to notice Muzzle no longer hiding notifications with Zoom.

@bdkjones it definitely worked with Zoom back in November (I was on Catalina or Catalina beta back then).

So either Zoom updated to change to some other screen sharing API, or this app needs updating to listen for Zoom events 🤔

**Comment by bdkjones** [comment]
There aren't any "zoom events" that the app listens for— Muzzle is hooked into a low-level part of the Mac's graphics system that triggers when the screensharing API starts piping data to, well, anything other than the normal display.

It is possible Zoom has been updated with a different approach; I'll have to take a look on a throwaway Mac. I refuse to install Zoom on my primary machine.

**Comment by calaway** [comment]
From Zoom's [changelog](https://support.zoom.us/hc/en-us/articles/201361963-New-Updates-for-macOS):
> August 25, 2019 Version 4.5.3261.0825
__Do Not Disturb Automatically Enabled when Screensharing__
When a user starts to share their screen, the operating system’s Do Not Disturb mode will be automatically enabled, muting all notifications.

As others have said, Muzzle was also working for me with Zoom and seems to have stopped recently. I did manually check for and installed Zoom updates (it was not an automatic update) last week, so I'm guessing this new version is what changed things.

I was getting mixed results after that, with it working intermittently. My best guess is that Muzzle and Zoom were stepping on each other's toes and both attempting to toggle Do Not Disturb, so it didn't work.

For now I just stopped Muzzle and Zoom is toggling Do Not Disturb correctly without it.

Either way, thank you for the awesome product! I have found it very useful up until now, and I suspect I will again in the future.

**Comment by bdkjones** [comment]
I suspect that’s the case. I’d be careful with Zoom, though. Very shoddy company, very shady engineering. When I’m forced to use Zoom, I do so on an ancient Mac that has no data on it. I would NEVER grant this app permissions on my production Mac.

**Comment by jspiro** [comment]
Yeah, I think the resolution here is to stop using Muzzle with Zoom. I found that my notifications were still hidden after ending screen sharing, probably because both apps toggled it.

**Comment by jspiro** [comment]
(Or to @bdkjones point, to stop using Zoom, if you can ;-))

**Comment by dseravalli** [comment]
Is there a way to add a check for a Zoom version or something? I use Google Meets and Zoom so it would be nice if Muzzle could play nice with Zoom.

**Comment by bdkjones** [comment]
Muzzle DOES work with Zoom. One of two things is going on:

1. The latest versions of Zoom have an option to automatically turn on DND when screensharing starts. If you use Muzzle, turn that option off so there is no conflict.

2. Muzzle only turns on DND when you SHARE YOUR SCREEN. Merely starting a video call won’t trigger Muzzle because nobody can see your notifications.

**Comment by glrodasz** [comment]
I followed the @bdkjones recommendations but I still saw notifications in zoom when I was sharing screen.

**Comment by karlhorky** [comment]
I also had this issue just now on Big Sur (macOS 11.1) (with Muzzle 1.7 and Zoom Version: 5.4.9 (59931.0110)).

Opened https://github.com/bdkjones/MuzzleApp/issues/90 for this in case you want to follow along @glrodasz  and @thegonch

**Comment by glrodasz** [comment]
I forgot to update the way to reproduce it, but apparently, when you toggle the shared screen in Zoom, it toggles the DND mode, only when Muzzle is opened.

**Comment by karlhorky** [comment]
> when you toggle the shared screen in Zoom, it toggles the DND mode, only when Muzzle is opened.

Hm... maybe I'm not understanding what you're saying here, but this almost sounds like you're saying that the behavior is... working correctly?

**Comment by glrodasz** [comment]
> > when you toggle the shared screen in Zoom, it toggles the DND mode, only when Muzzle is opened.
> 
> Hm... maybe I'm not understanding what you're saying here, but this almost sounds like you're saying that the behavior is... working correctly?

Right, I just tested and this is what is happening:

1st time: I shared screen, Muzzle enabled DND, I stop sharing, DND keeps enabled.
2nd time: I shared screen DND keeps enabled, I stop sharing, DND is disabled.
3rd time: I shared screen, Muzzle enabled DND, I stop sharing, DND keeps enabled.

Before when you started the 2nd time since DND was enabled, after stop sharing, the start sharing screen was disabling DND.
